pattern/variable/ThingVariable.java

```diff
@@ -249,14 +250,10 @@ public ThingVariable asThing() {
     }
 
     @Override
-    public AlphaEquivalence alphaEquals(ThingVariable that) {
-        return AlphaEquivalence.valid()
-                .validIf(id().isName() == that.id().isName())
-                .validIf(this.inferredTypes().equals(that.inferredTypes()))
-                .validIfAlphaEqual(this.isaConstraint, that.isaConstraint)
-                .validIfAlphaEqual(this.relationConstraint, that.relationConstraint)
-                .validIfAlphaEqual(this.hasConstraints, that.hasConstraints)
-                .validIfAlphaEqual(this.valueConstraints, that.valueConstraints)
-                .addMapping(this, that);
+    public FunctionalIterator<AlphaEquivalence> alphaEquals(ThingVariable that) {
+        return AlphaEquivalence.empty()
+                .alphaEqualIf(id().isName() == that.id().isName())
+                .flatMap(a -> a.alphaEqualIf(this.inferredTypes().equals(that.inferredTypes())))
+                .map(a -> a.extend(this, that));
```

Comment:
here we encode that assumption that alpha equivalence only works over named variables -- we can have that encoded here, but the `extend()` method maybe should actually be able to take any type of variable and handle it?

ie. maybe we should get rid of the assertion that variables can only have named IDs when extending the alpha equivalence

Reply:
Again, same as above, this check makes sure the variables are either both named or both anonymous, so there is no change to make.